src/acl/mock.ts

```diff
@@ -0,0 +1,123 @@
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import {
+  WithResourceInfo,
+  WithResourceAcl,
+  WithAcl,
+  WithAccessibleAcl,
+  WithFallbackAcl,
+  UrlString,
+} from "../interfaces";
+import { getSourceIri } from "../resource/resource";
+import { createAcl, internal_getContainerPath } from "./acl";
+import { mockContainerFrom } from "../resource/mock";
+
+/**
+ * Function for use in unit tests to mock a [[SolidDataset]] that has its own ACL attached.
+ *
+ * Warning: do not use this function in actual production code.
+ * This function initialises a new empty ACL and attaches it to a given [[SolidDataset].
+ * This is useful to mock a SolidDataset with an ACL in tests of code that call e.g.
+ * [[getResourceAcl]].
+ *
+ * @param resource The Resource that we should pretend has its own ACL.
+ * @returns The input Resource, with an empty Resource ACL attached.
+ * @since Not released yet.
+ */
+export function addMockResourceAclTo<T extends WithResourceInfo>(
+  resource: T
+): T & WithResourceAcl {
+  const aclUrl =
+    resource.internal_resourceInfo.aclUrl ?? getSourceIri(resource) + ".acl";
+  const resourceWithAclUrl: typeof resource & WithAccessibleAcl = Object.assign(
+    resource,
+    {
+      internal_resourceInfo: {
+        ...resource.internal_resourceInfo,
+        aclUrl: aclUrl,
+      },
+    }
+  );
+  const aclDataset = createAcl(resourceWithAclUrl);
+
+  const resourceWithResourceAcl: typeof resourceWithAclUrl &
+    WithResourceAcl = Object.assign(resourceWithAclUrl, {
+    internal_acl: {
+      resourceAcl: aclDataset,
+      fallbackAcl:
+        ((resourceWithAclUrl as unknown) as WithAcl).internal_acl
+          ?.fallbackAcl ?? null,
+    },
+  });
+
+  return resourceWithResourceAcl;
+}
+
+/**
+ * Function for use in unit tests to mock a [[SolidDataset]] that has one of its Container's ACL attached.
+ *
+ * Warning: do not use this function in actual production code.
+ * This function initialises a new empty ACL and attaches it to a given [[SolidDataset] as its
+ * fallback ACL.
+ * This is useful to mock a SolidDataset with an ACL in tests of code that call e.g.
+ * [[getFallbackAcl]].
+ *
+ * @param resource The Resource that we should pretend has one of its Container's ACL attached.
+ * @returns The input Resource, with an empty Fallback ACL attached.
+ * @since Not released yet.
+ */
+export function addMockFallbackAclTo<T extends WithResourceInfo>(
+  resource: T
+): T & WithFallbackAcl {
+  const containerUrl = internal_getContainerPath(getSourceIri(resource));
+  const aclUrl = containerUrl + ".acl";
+  const mockContainer = setMockAclUrl(mockContainerFrom(containerUrl), aclUrl);
+  const aclDataset = createAcl(mockContainer);
+
+  const resourceWithFallbackAcl: typeof resource &
+    WithFallbackAcl = Object.assign(resource, {
+    internal_acl: {
+      resourceAcl:
+        ((resource as unknown) as WithAcl).internal_acl?.resourceAcl ?? null,
```

Comment:
For my own education, why is it necessary to cast to unknown before casting to `WithAcl` ?

Reply:
Because the type of `resource` is not assignable to `WithAcl`, and vice versa. The background knowledge here is that these are not casts, but type assertions, i.e. they do not actually change the type of the object, but tell TypeScript that you know better than it what the type is. Which is often fine if the types are somewhat compatible, but since it's more likely to not be what you meant if they're completely complementary, TypeScript does warn you when you try to do that. And since `unknown` is compatible with all types, the double assertion does work.

It's explained more clearly here: https://basarat.gitbook.io/typescript/type-system/type-assertion#double-assertion

Reply:
Thanks !